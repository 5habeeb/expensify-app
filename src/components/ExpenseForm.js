import React from 'react';
import moment from 'moment'; // moment is the stadard for dates, always use it
import { SingleDatePicker } from 'react-dates';

class ExpenseForm extends React.Component {
    // constructor(props) {
    //     super(props);
    //     this.state = {
    //         description: props.expense ? props.expense.description : '',
    //         note: props.expense ? props.expense.note : '',
    //         amount: props.expense ? (props.expense.amount / 100).toString() : '',
    //         createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
    //         calendarFocused: false, 
    //         error: ''
    //     }
    // }

    state = {
        description: this.props.expense ? this.props.expense.description : '',
        note: this.props.expense ? this.props.expense.note : '',
        amount: this.props.expense ? (this.props.expense.amount / 100).toString() : '',
        createdAt: this.props.expense ? moment(this.props.expense.createdAt) : moment(),
        calendarFocused: false, 
        error: ''
    }

    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({ description }));
    };

    OnNoteChange = (e) => {
        //another way of using event
        //e.persist();
        //this.setState(() => ({ note: e.target.value }));
        const note = e.target.value;
        console.log(note);
        this.setState(() => ({ note }));
    };

    onAmountChange = (e) => {
        const amount = e.target.value;
        console.log(amount);
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount }));
        }
    };

    onDateChange = (createdAt) => {
        if (createdAt){
            this.setState(() => ({ createdAt }));
        }
    };

    onFocusChange = ({ focused }) => {
        this.setState(() => ({ calendarFocused: focused }));
    };

    onSubmit = (e) => {
        e.preventDefault(); // prevent page refresh

        if(!this.state.description || !this.state.amount){
            this.setState(() => ({ error: 'Please add a description and an amount!' }));
        } else {
            this.setState(() => ({ error: '' }));
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100,
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            });
        }
    };

    render() {
        return (
            <div>
                {this.state.error && <p>{this.state.error}</p>}
                <form onSubmit={this.onSubmit}>
                    <input 
                        type="text"
                        placeholder="Description"
                        autoFocus={true}
                        value={this.state.description}
                        onChange={this.onDescriptionChange}
                    />
                    <input 
                        type="text"
                        placeholder="Amount"
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                    />
                    <SingleDatePicker 
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={(day) => false }
                    />
                    <textarea 
                        placeholder="Add a note for your expense (optional)"
                        cols="30" 
                        rows="10"
                        onChange={this.OnNoteChange}
                        value={this.state.note}
                        >
                    </textarea>
                    {this.props.expense ? <button>Save Expense</button> : <button>Add Expense</button>}
                    
                </form>
            </div>
        );
    }
}

export default ExpenseForm;