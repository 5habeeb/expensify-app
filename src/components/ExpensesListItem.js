import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral'

const ExpensesListItem = ({ description, amount, createdAt, id }) => (
    <div>
        <Link to={`/edit/${id}`}>
            <h3>{description}</h3>
        </Link>
        <p>
            {numeral(amount / 100).format('$0,00.00')}
            {' - '} 
            {moment(createdAt).format('MMMM Do, YYYY')}
        </p>
    </div>
);
export default ExpensesListItem;

// Any components which may communicate with the store
// should be connected to the store
