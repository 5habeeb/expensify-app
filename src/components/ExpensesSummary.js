import React from 'react';
import { connect } from 'react-redux';
import selectExpenses from '../selectors/Expenses';
import getExpensesTotal from '../selectors/expenses-total';
import numeral from 'numeral';

export const ExpensesSummary = ({ expenseCount, expensesTotal }) =>  {
    const expenseWord = (expenseCount === 1) ? 'expense' : 'expenses'; 
    const formattedExpensesTotal = numeral(expensesTotal / 100).format('$0,00.00');
    return (
        <div>
        {
            expenseCount ? (
                <p>{expenseCount} {expenseWord} of total {formattedExpensesTotal}</p>
            ) : (
                <p>Expenses total {formattedExpensesTotal}</p>
            )
        }
        </div>
    );
};

const mapStateToProps = (state) => {
    const visiblieExpenses = selectExpenses(state.expenses , state.filters);
    return {
        expenseCount: visiblieExpenses.length,
        expensesTotal: getExpensesTotal(visiblieExpenses)
    }
};

export default connect(mapStateToProps)(ExpensesSummary);