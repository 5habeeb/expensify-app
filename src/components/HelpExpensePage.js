import React from 'react';

const HelpExpensePage = () => (
    <div>
        this is from my help component
    </div>
);

export default HelpExpensePage;