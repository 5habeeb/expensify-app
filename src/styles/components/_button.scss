.button {
    background: $blue;
    color: white;
    border: none;
    border-radius: 5px;
    padding: $m-size;
    font-size: $font-size-large;
    font-weight: 300;
    display: inline-block;
    text-decoration: none;
    line-height: 1;
}

.button--link {
    background: none;
}

.button--remove {
    background: #888888;
}