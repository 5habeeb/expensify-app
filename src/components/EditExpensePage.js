import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { editExpense, startRemoveExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component{
    onSubmit = (expense) => {
        this.props.editExpense(this.props.expense.id, expense);
        this.props.history.push('/');
    };

    handelClick = (e) => {
        this.props.startRemoveExpense({ id: this.props.expense.id });
        this.props.history.push('/');
        }

    render() {
        return (
            <div>
                <ExpenseForm
                    expense={this.props.expense}
                    onSubmit={this.onSubmit}
                />
                <button onClick={this.handelClick}
            > Remove expense</button>
            </div>
        );
    };
};

const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find( (expense) => props.match.params.id === expense.id )
    };
};

const mapDispatchTpProps = (dispatch) => ({ 
    editExpense: (id, expense) => dispatch(editExpense(id, expense)),
    startRemoveExpense: (data) => dispatch(startRemoveExpense(data))
});

export default connect(mapStateToProps, mapDispatchTpProps)(EditExpensePage);