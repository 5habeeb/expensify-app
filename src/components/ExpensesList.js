import React from 'react';
import { connect } from 'react-redux';
import ExpensesListItem from './ExpensesListItem';
import selectExpenses from '../selectors/Expenses';

export const ExpensesList = (props) => (
    <div className="content-container">
        <div className="list-header" >
            <div className="show-for-mobil">Expenses</div>
            <div className="show-for-desktop">Expense</div>
            <div className="show-for-desktop">Amount</div>
        </div>
        <div className="list-body">
            {
                props.expenses.length === 0 ? (
                    <div className="list-item--message">
                        <span>No Expenses</span>
                    </div>
                ) : (
                        props.expenses.map ( (expense) => (
                            <ExpensesListItem key={expense.id} {...expense}/>
                            )   
                        )
                    )
            }
        </div>
    </div>
);

const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses , state.filters)
    };
};

export default connect(mapStateToProps)(ExpensesList);