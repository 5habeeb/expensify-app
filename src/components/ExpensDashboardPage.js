import React from 'react';
import ExpensesList from './ExpensesList';
import ExpensesListFilters from './ExpensesListFilters';

const ExpensDashboardPage = () => (
    <div>
        <ExpensesListFilters />
        <ExpensesList />
    </div>
);

export default ExpensDashboardPage;