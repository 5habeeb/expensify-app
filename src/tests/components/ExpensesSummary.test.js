import React from 'react';
import { shallow } from 'enzyme';
import { ExpensesSummary } from '../../components/ExpensesSummary';
import expenses from '../fixtures/expenses';
import getExpensesTotal from '../../selectors/expenses-total';

test('should render Expenses summary with no expenses', ()=> {
    const wrapper = shallow(<ExpensesSummary expenses={[]} expenseCount={0} />);
    expect(wrapper).toMatchSnapshot();
});

test('should render Expenses summary with multipe expenses', ()=> {
    const expensesTotal = getExpensesTotal(expenses);
    const wrapper = shallow(<ExpensesSummary expensesTotal={expensesTotal} expenseCount={expenses.length} />);
    expect(wrapper).toMatchSnapshot();
});