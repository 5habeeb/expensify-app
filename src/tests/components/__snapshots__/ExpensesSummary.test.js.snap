// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Expenses summary with multipe expenses 1`] = `
<div>
  <p>
    You have 
    3
     
    expenses
     of total 
    $1,141.95
  </p>
</div>
`;

exports[`should render Expenses summary with no expenses 1`] = `
<div>
  <p>
    You have no expenses
  </p>
</div>
`;

exports[`should render Expenses summary with single expenses 1`] = `
<div>
  <p>
    You have 
    1
     
    expense
     of total 
    $45.00
  </p>
</div>
`;
